src/nftfeed.sol

```diff
@@ -0,0 +1,160 @@
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) external auth {
+        if (contractName == "pile") {pile = PileLike(addr);}
+        else if (contractName == "shelf") { shelf = ShelfLike(addr); }
+        else revert();
+    }
+
+    /// returns a unique id based on registry and tokenId
+    /// the nftID allows to define a risk group and an nft value
+    /// before a loan is issued
+    function nftID(address registry, uint tokenId) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(registry, tokenId));
+    }
+
+    function nftID(uint loan) public view returns (bytes32) {
+        (address registry, uint tokenId) = shelf.shelf(loan);
+        return nftID(registry, tokenId);
+    }
+
+    /// Admin -- Updates
+    function setRiskGroup(uint risk_, uint thresholdRatio_, uint ceilingRatio_, uint rate_) internal {
+        thresholdRatio[risk_] = thresholdRatio_;
+        ceilingRatio[risk_] = ceilingRatio_;
+        // the risk group is used as a rate id in the pile
+        pile.file("rate", risk_, rate_);
+    }
+
+    ///  -- Oracle Updates --
+
+    /// update the nft value
+    function update(bytes32 nftID_,  uint value) public auth {
+        nftValues[nftID_] = value;
+    }
+
+    /// update the nft value and change the risk group
+    function update(bytes32 nftID_, uint value, uint risk_) public auth {
+        require(thresholdRatio[risk_] != 0, "threshold for risk group not defined");
+
+        // change to new rate immediately in pile if a loan debt exists
+        uint loan = shelf.nftlookup(nftID_);
+        if (pile.pie(loan) != 0) {
+            pile.changeRate(loan, risk_);
+        }
+
+        risk[nftID_] = risk_;
+        nftValues[nftID_] = value;
+    }
+
+    // method is called by the pile to check the ceiling
+    function borrow(uint loan, uint amount) external auth {
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+    }
+
+    // method is called by the pile to check the ceiling
+    function repay(uint loan, uint amount) external auth {}
+
+    // borrowEvent is called by the shelf in the borrow method
+    function borrowEvent(uint loan) public auth {
+        uint risk_ = risk[nftID(loan)];
+
+        // condition is only true if there is no outstanding debt
+        // if the rate has been changed with the update method
+        // the pile rate is already up to date
+        if(pile.loanRates(loan) != risk_) {
+            pile.setRate(loan, risk_);
+        }
+    }
+
+    // unlockEvent is called by the shelf.unlock method
+    function unlockEvent(uint loan) public auth {}
+
+    ///  -- Getter methods --
+    /// returns the ceiling of a loan
+    /// the ceiling defines the maximum amount which can be borrowed
+    function ceiling(uint loan) public view returns (uint) {
```

Comment:
our current implementation of the ceiling interface returns your current ceiling, considering the current debt

Reply:
I changed it to principal implementation